op-chain-ops/script/check_super_root.go

```diff
@@ -0,0 +1,328 @@
+	}
+
+	if err := m.calculateTargetBlockNumbers(ctx); err != nil {
+		return fmt.Errorf("failed to calculate target block numbers: %w", err)
+	}
+
+	if err := m.findAnchorTimestamp(ctx); err != nil {
+		return fmt.Errorf("failed to find anchor timestamp: %w", err)
+	}
+	m.log.Info("Found common anchor timestamp", "timestamp", m.anchorTimestamp)
+
+	if err := m.calculateOutputRoots(ctx); err != nil {
+		return fmt.Errorf("failed to calculate output roots: %w", err)
+	}
+
+	if err := m.calculateSuperRoot(); err != nil {
+		return fmt.Errorf("failed to calculate super root: %w", err)
+	}
+
+	m.log.Info("Super root calculation process finished successfully")
+	return nil
+}
+
+// initClientsAndFetchIDs establishes connections to all RPC endpoints
+// and retrieves their chain IDs.
+func (m *SuperRootMigrator) initClientsAndFetchIDs(ctx context.Context) error {
+	m.log.Info("Initializing clients and fetching chain IDs...")
+
+	for _, endpoint := range m.rpcEndpoints {
+		m.log.Debug("Dialing RPC endpoint", "url", endpoint)
+		client, err := ethclient.DialContext(ctx, endpoint)
+		if err != nil {
+			return fmt.Errorf("failed to connect to RPC endpoint %s: %w", endpoint, err)
+		}
+		m.ethClients[endpoint] = client
+		m.log.Info("Connected to client", "url", endpoint)
+
+		chainID, err := client.ChainID(ctx)
+		if err != nil {
+			// Clean up the client we just created before returning
+			client.Close()
+			delete(m.ethClients, endpoint)
+			return fmt.Errorf("failed to get chain ID from %s: %w", endpoint, err)
+		}
+		m.log.Info("Fetched chain ID", "url", endpoint, "chain_id", chainID)
+
+		m.chainSettings[endpoint] = &ChainSettings{
+			ChainID: chainID,
+			RPCURL:  endpoint,
+		}
+	}
+
+	m.log.Info("Successfully initialized clients", "count", len(m.ethClients))
+	return nil
+}
+
+func (m *SuperRootMigrator) calculateTargetBlockNumbers(ctx context.Context) error {
+	m.log.Info("Calculating target block numbers...")
+
+	for endpoint, client := range m.ethClients {
+		// Get the latest block
+		latestBlock, err := client.BlockByNumber(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("failed to get latest block from %s: %w", endpoint, err)
+		}
+
+		// Get the parent block
+		parentBlock, err := client.BlockByNumber(ctx, big.NewInt(latestBlock.Number().Int64()-1))
+		if err != nil {
+			return fmt.Errorf("failed to get parent block from %s: %w", endpoint, err)
+		}
+
+		// Calculate block time (difference in timestamps between latest and parent blocks)
+		blockTime := latestBlock.Time() - parentBlock.Time()
+		if blockTime == 0 {
+			return fmt.Errorf("block time cannot be zero for chain %s", endpoint)
+		}
+
+		// Calculate how many blocks to look back to reach the target timestamp
+		timeDiff := latestBlock.Time() - m.TargetTimestamp
+		blocksToLookBack := (timeDiff / blockTime) + 1
+
+		// Compute the target block number
+		targetBlockNumber := big.NewInt(latestBlock.Number().Int64() - int64(blocksToLookBack))
```

Comment:
We should make sure the timestamp isn't before genesis and can simplify the subtraction slightly.

```suggestion

		if blocksToLookBack > latestBlock.NumberU64() {
			return fmt.Errorf("target timestamp is prior to genesis for endpoint %v", endpoint)
		}
		// Compute the target block number
		targetBlockNumber := new(big.Int).SetUint64(latestBlock.Number().Uint64() - blocksToLookBack)
```